3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2017-2018 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package com.snowplowanalytics
+package snowplow.enrich
+package common.enrichments.registry
+
+// Scala
+import scala.collection.JavaConverters._
+
+// Snowplow
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.jayway.jsonpath.MapFunction
+import com.snowplowanalytics.iglu.client.{SchemaCriterion, SchemaKey}
+import com.snowplowanalytics.snowplow.enrich.common.ValidatedNelMessage
+import com.snowplowanalytics.snowplow.enrich.common.utils.MapTransformer.TransformMap
+import com.snowplowanalytics.snowplow.enrich.common.utils.ScalazJson4sUtils
+
+// Iglu
+import iglu.client.validation.ProcessingMessageMethods._
+
+// Scala libraries
+import org.json4s.JValue
+import org.json4s.JsonAST._
+import org.json4s.jackson.JsonMethods
+import org.json4s.jackson.JsonMethods.{compact, parse, render}
+import org.json4s.DefaultFormats
+import org.json4s.MappingException
+
+// Java
+import java.security.{MessageDigest, NoSuchAlgorithmException}
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.{ObjectNode, TextNode}
+import com.jayway.jsonpath.spi.json.JacksonJsonNodeJsonProvider
+import com.jayway.jsonpath.{Configuration, JsonPath => JJsonPath, Option => JOption}
+
+// Scalaz
+import scalaz._
+import Scalaz._
+
+/**
+ * PiiField trait. This corresponds to a configuration top-level field (i.e. either a POJO or a JSON field) along with
+ * a function to apply that strategy to a TransformMap
+ *
+ * Members are:
+ * strategy: the strategy to be used in this field
+ * fieldName: the fieldName where this startegy will be applied
+ * transformer: a function that modifies the TransformMap by adding PII transformations where appropriate
+ */
+sealed trait PiiField {
+  def strategy: PiiStrategy
+  def fieldName: String
+  protected def applyStrategy(fieldValue: String): String
+  def transformer(transformMap: TransformMap) =
+    transformMap.collect {
+      case (inputField: String, (tf: Function2[String, String, Validation[String, String]], outputField: String))
```

Comment:
did we solve this type erasure issue?

Reply:
How should we solve it? Should we add a ClassTag?

Reply:
The good way to start would be to:

1. Understand how it impacts `TransformMap`'s fields like `vp`, `res` and others if they return something beside of plain `String`
2. Make it very-very clear in pattern-matching that it does not imply types that are written there now
3. Write unit-tests with those fields
4. Submit an issue, reflecting that we need to refactor `TransformMap`. This is probably going to be part of much bigger atomic events refactoring, but still worth to note that type-safety problem affects PII Enrichment
5. Depending on impact of problem - we'll probably have to define overloaded `applyStrategy` accepting `(String, String)` and pattern-match it inside `tf` or leave it full of warnings:

```scala
f(k, v).map {
  case s: String => applyStrategyString(s)
  case p: (String, String) => applyStrategyPair(p)
}
```
6. Make it more readable to so that peers wouldn't have to spend time to parse and re-write it into more readable format

What is more important that this question should have been raised before claiming that all issues were addressed.